pep-9999.txt

```diff
@@ -0,0 +1,171 @@
+PEP: 9999
+Title: Python local packages directory [We need to find a proper name]
+Version: $Revision$
+Last-Modified: $Date$
+Author: Kushal Das <mail@kushaldas.in>, Steve Dower <steve.dower@python.org>,
+        Donald Stufft <donald@stufft.io>, Nick Coghlan <ncoghlan@gmail.com>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 16-May-2018
+Python-Version: 3.8
+
+
+
+Abstract
+========
+
+This PEP proposes to add to Python a mechanism to have a ``__pypackages__``
+directory  per project to have its own independent set of installed packages.
+This will share the standard library with the base installed Python. This will
+avoid the steps to create or activate or deactivate "virtual environments". Any
+Python scripts will use the ``__pypackages__`` from the base directory of the 
+script.
+
+
+
+Motivation
+==========
+
+Python virtual environments have become an essential part of development and
+teaching workflow in the community, but at the same time, they create a
+barrier to entry for many folks. The following are few examples of the issues
+people get into while being introduced to Python (or programming for the first
+time).
+
+- How virtual environments work is a lot of information for anyone new. It takes
+  a lot of extra time and effort to explain them.
+
+- Different platforms and shell environments require different sets of commands
+  to work on the virtual environments. Any workshop or teaching environment with
+  people coming with different operating systems installed on the laptops create a
+  lot of confusion among the participants.
+
+- Virtual environments need to be activated on each opened terminal, if someone
+  creates/opens a new terminal, that by default does not get the same environment
+  as in a previous terminal with virtual environment activated.
+
+
+Specification
+=============
+
+When the Python binary is executed, it attempts to determine its
+prefix (which it stores in ``sys.prefix``), which is then used to find
+the standard library and other key files, and by the ``site`` module
+to determine the location of the site-package directories.  Currently
+the prefix is found (assuming ``PYTHONHOME`` is not set) by first
+walking up the filesystem tree looking for a marker file (``os.py``)
+that signifies the presence of the standard library, and if none is
+found, falling back to the build-time prefix hard coded in the binary.
+
+This PEP proposes to add a new step in this process. If a ``__pypackages__``
+directory is found in the current working directory, then it will be included in
+``sys.path`` after the current working directory and just before the system
+site-packages. This way, if the Python executable starts in the given project
```

Comment:
We might want to think about if we want to include the system ``site-packages`` at all or not. Virtualenv has a configuration that lets you choose which it is, but the default has, for a long time now, been to isolate from the system ``site-packages``, and I think that is generally a good default. However it is also useful to sometimes be able to use the system ``site-packages``, so is there some way we can make this configurable? Perhaps a config file inside of ``__pypackages__`` or maybe just a marker file that indicates which it should do.

If no configuration is possible, I'd argue that it'd be better to always isolate rather than never isolate. We can use symlinks or something to pull in a system level installed item at the very worst.

Reply:
We also probably want to specify where user site packages go in this order too (local, user, global IMO).

Reply:
Having no configuration seems reasonable to me as well. If we go the no configuration approach then I think it should not include ``--user`` or ``--global`` in ``sys.path`` if it locates one of these directories. Overlaying is how virtualenv used to work, and we explicitly moved away from it because it caused a lot of confusion and issues for people.

With regards ot if pip ignoes user/global when resolving, *if* we ignore them, then you lose most of the useful-ness of having them on ``sys.path`` to begin with since the primary reason you'd want that is to be able to use a system lib instead of a lib from PyPI, but if we're ignoring them we're going to resolve and install the PyPI version.

Reply:
> The other option is to simply tell folks that if they want site-packages they have to use a virtual environment or some other mechanism. If we're doing this for simplicity then we should think about keeping it simple.

But, if I have activated virtualenv or another mechanism, would that *override* `__pypackages__`?

If a project distributes an empty `__pypackages__`, as recommended below, does that mean system's copy of `requests` is simply unusable with that project?
I would not like that.

Reply:
To be clear, ``__pypackages__`` is not something that a project would distribute through something like PyPI or similiar. It's more for the case of "git clone, and then install these dependencies". If you didn't want to use it, you'd just delete the ``__pypackages__``, (but generally, I don't think projects should be committing empty ``__pypackages__`` anyways).

Reply:
> Overlaying is how virtualenv used to work, and we explicitly moved away from it because it caused a lot of confusion and issues for people.

Isolation is a good default for a virtual environment, but this is more light-weight.

Would an empty `__pypackages__` directory mean `pip` is not available?

Reply:
If isolated by default, it'd mean that ``python -m pip`` is unavailable, because ``python -m`` would look in CWD and find ``__pypackages__``. It would not affect ``pip`` because pip would be located at ``/usr/bin/pip`` or similar and we wouldn't resolve starting at CWD for a ``__pypackages__``.

Unrelated to this PEP, I've been torn on how to resolve the many ways of invoking pip and specifying what your target environment is. I think there have been two realistic outcomes to that, which is either try to focus on ``python -m pip`` and letting the normal ``$PATH`` resolution for ``python`` manage what target environment to use or focus on ``pip`` and add an option to point to a specific environment (e.g. ``pip install --with-python /usr/bin/python3``). There are positives and negatives in both directions and it's not really a question for this particular PEP, but I thought it might be important to mention at least!

Reply:
Having ```__pypackages__``` overlay with site-packages and user-packages seems like the best thing for the audience of teaching new users with the least fuss and bother. If you exclude packages from the system, then users will end up having to learn two systems in order to get things done (pip to install some packages into the __pypackages__ directory and ln -s to install other ones into there.)  This also has the benefit of expanding on the already present concept of having a search path for python libraries so it's only one concept to learn.

Reply:
If they overlay, would a package installer be expected to consider the dependencies from an overlaid location in the dependency resolution? E.g. if I ``apt-get install python-requests && pip install --local thing-that-depends-on-requests`` does pip see that requests is already satisfied and skip that particular dependency, or does it install requests again into ``__pypackages__``?

Reply:
Yes, I'd expect `pip` to skip installing requests in that case. (Which does bring trouble if requests is then uninstalled, though...)

Reply:
For what it's worth, after looking closer at what node.js does here (since they're the prior art in this space) they do overlay the local ``node_modules/`` directory with their form of user and global site-packages. So it appears that the prior art in this space is *not* isolating them by default.

Reply:
If something is found in user or global `site-packages` as a dependency at install-time then you will potentially break your packages if you update them at all which will also trip up new users. So if some user went home, found out about `python3 -m pip install --upgrade` and it failed because of something in their global or user install then it might be hard for them to diagnose.

I also don't understand what @abadger means by having to teach new users `ln -s` as a second way to install anything? If you were leading e.g. a one-day class, I would  assume you would say, "OK everyone, run `pip install --local -r requirements.txt` to get your dependencies" and that's it.

Reply:
> If they overlay, would a package installer be expected to consider the dependencies from an overlaid location in the dependency resolution?


Absolutely not. If you install a package with `__pypackages__` there, you should get the package and all its dependencies installed into the directory, not some random mix of packages. But if we don't overlay, `python -m pip ...` cannot work.

I think the only odd case is `pip install --local --upgrade (something installed in --global)`, which *could* be expected to upgrade the `--global` install rather than installing new into `--local`. But perhaps printing a warning in this case is enough?